src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
+            sub_comb = self._tensordot(
+                reduced,
+                trace_out,
+                dims=([0], [0]),
+            )
+            expected = self._tensordot(trace_out / dim_out, sub_comb, dims=0)
+        else:
+            reduced = self._tensordot(self.full(), trace_in, axes=(0, 0))
+            sub_comb = self._tensordot(reduced, trace_out, axes=(0, 0))
+            expected = self._tensordot(trace_out / dim_out, sub_comb, axes=0)
+
+        norm = self._backend.calculate_norm((reduced - expected), order=order)
+        if float(norm) > precision_tol:
+            return False
+        elif len(self.partition) == 2:
+            return True
+        else:
+            # Unital is defined for quantum channels only. But we can extend it to quantum combs as follows:
+            return QuantumChannel(  # pragma: no cover
+                sub_comb, self.partition[2:], pure=False, backend=self._backend
+            ).is_unital(order, precision_tol)
+
+    def is_channel(
+        self,
+        order: Optional[Union[int, str]] = None,
+        precision_tol_causal: float = 1e-8,
+        precision_tol_psd: float = 1e-8,
+    ):
+        """Returns bool indicating if Choi operator :math:`\\mathcal{E}` is a channel.
+
+        Args:
+            order (int or str, optional): order of the norm used to calculate causality.
+                Defaults to ``None``.
+            precision_tol_causal (float, optional): threshold :math:`\\epsilon` that defines if
+                Choi operator of the network is :math:`\\epsilon`-close to causality in the norm
+                given by ``order``. Defaults to :math:`10^{-8}`.
+            precision_tol_psd (float, optional): threshold value used to check if eigenvalues of
+                the Choi operator :math:`\\mathcal{E}` are such that
+                :math:`\\textup{eigenvalues}(\\mathcal{E}) >= \\textup{precision_tol_psd}`.
+                Note that this parameter can be set to negative values.
+                Defaults to :math:`0.0`.
+
+        Returns:
+            bool: Channel condition.
+        """
+        return self.is_causal(
+            order, precision_tol_causal
+        ) and self.is_positive_semidefinite(precision_tol_psd)
+
+    def apply(self, state):
+        """Apply the Choi operator :math:`\\mathcal{E}` to ``state`` :math:`\\varrho`.
+
+        It is assumed that ``state`` :math:`\\varrho` is a density matrix.
+
+        Args:
+            state (ndarray): density matrix of a ``state``.
+
+        Returns:
+            ndarray: Resulting state :math:`\\mathcal{E}(\\varrho)`.
+        """
+        operator = self.copy().operator()
+        conj = self._backend.np.conj
 
         if self.is_pure():
-            matrix = self._einsum("jk,lm -> kjml", matrix, np.conj(matrix))
+            return self._einsum("ij,lk,il", operator, conj(operator), state)
+
+        return self._einsum("ijkl, jl", operator, state)
+
+
+class StochQuantumNetwork:
+    pass
```

Comment:
?

Reply:
Stochastic Quantum Networks are planned but not to be implemented in this PR. Do we need to delete it here?

Reply:
Yes, it's best to delete from this PR and in the future create a separate PR for it